### PR TITLE
Support Extra Propagation Fields with Trace

### PR DESCRIPTION
**Pull request by meltsufin**
Providing a Trace Propagation Factory Builder instead of just the Factory,
which allows us to rely on the `TraceAutoConfiguration.sleuthPropagation()`
rather than completely skipping it.

Fixes: #2268.

### CODE REVIEW

**Comment by meltsufin** [comment]
@adriancole Ultimately, what we're after is the ability to configure the delegate for `TraceAutoConfiguration`, which is currently hardcoded as `B3Propagation.FACTORY`. I'm open to whatever would be the best way to do it. It will probably require some tweaks to `TraceAutoConfiguration` in Sleuth though.

**Comment by adriancole** [comment]
the exception will go away with a new brave release which will immediately go into the next sleuth release.

https://github.com/openzipkin/brave/pull/1128

**Comment by meltsufin** [comment]
@adriancole Thanks for such a quick turnaround!

File: spring-cloud-gcp-dependencies/pom.xml [hunk 1]
```diff
@@ -16,8 +16,8 @@
 	<packaging>pom</packaging>
 
 	<properties>
-		<gcp-libraries-bom.version>4.4.1</gcp-libraries-bom.version>
-		<cloud-sql-socket-factory.version>1.0.15</cloud-sql-socket-factory.version>
+		<gcp-libraries-bom.version>5.3.0</gcp-libraries-bom.version>
```
Comment (elefeint) [inline spring-cloud-gcp-dependencies/pom.xml:19]: Does it make sense to move the version updates to a separate PR?

**Comment by meltsufin** [comment]
Looks like `TraceSampleApplicationTests` fails as soon as we upgrade `spring-cloud-sleuth` to `2.2.3.BUILD-SNAPSHOT`. 😞

**Comment by elefeint** [comment]
I can look into it Thursday -- I've had a lot of tracing on my mind lately.

**Comment by meltsufin** [comment]
We should rework this PR based on https://github.com/spring-cloud/spring-cloud-sleuth/pull/1647 as well.

**Comment by elefeint** [comment]
My first thought was that we needed a newer brave, but sleuth snapshot already pulls in the latest...

**Comment by elefeint** [comment]
Ah. `HexCodec` got moved into a subpackage in openzipkin/brave#1193.

We need to update https://github.com/openzipkin/zipkin-gcp/blob/8b2462c054737259bdd707de49da0cc32aaa6806/propagation-stackdriver/src/main/java/zipkin2/propagation/stackdriver/XCloudTraceContextExtractor.java

**Comment by meltsufin** [comment]
See: https://github.com/openzipkin/zipkin-gcp/issues/175

**Comment by adriancole** [comment]
zipkin-gcp 0.17.0 is on the way now. shading the internal HexCodec thing so it should be safe.

Notably, this fixes the propagation package also, though former is still there. brave.propagation instead of zipkin2.propagation

Finally, this allows you to control the underlying B3PropagationFactory, as mentioned 
https://github.com/spring-cloud/spring-cloud-sleuth/pull/1647

One subtle difference 2.2.x to 3.x is the change in the primary inject format.

2.2.x is
```java
return BaggagePropagation.newFactoryBuilder(B3Propagation.newFactoryBuilder()
                         .injectFormat(B3Propagation.Format.MULTI).build());
```

3.0 is
```java
return BaggagePropagation.newFactoryBuilder(B3Propagation.newFactoryBuilder()
                         .injectFormat(B3Propagation.Format.SINGLE_NO_PARENT).build());
```

File: spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/trace/StackdriverTraceAutoConfiguration.java [hunk 2]
```diff
@@ -221,8 +218,8 @@ public Sender stackdriverSender(GcpTraceProperties traceProperties,
 
 	@Bean
 	@ConditionalOnMissingBean
-	public Propagation.Factory stackdriverPropagation() {
-		return StackdriverTracePropagation.FACTORY;
+	public BaggagePropagation.FactoryBuilder baggagePropagationFactoryBuilder() {
+		return BaggagePropagation.newFactoryBuilder(StackdriverTracePropagation.FACTORY);
```
Comment (adriancole) [inline spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/trace/StackdriverTraceAutoConfiguration.java:222]: right here will change with 0.17 we should use the correct brave.propagation package and add a comment that the underlying B3Propagation.Factory needs to be configured differently for sleuth 3.0

This can be a TODO I suppose as 3.0 isn't final anyway

File: spring-cloud-gcp-samples/spring-cloud-gcp-trace-sample/src/test/java/brave/internal/HexCodec.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package brave.internal;
+
+// code originally imported from zipkin.Util
+public final class HexCodec {
```
Comment (adriancole) [inline spring-cloud-gcp-samples/spring-cloud-gcp-trace-sample/src/test/java/brave/internal/HexCodec.java:17]: as this is src/test ... maybe just use Long.whatever? HexCodec is mostly for performance

Comment (meltsufin) [reply spring-cloud-gcp-samples/spring-cloud-gcp-trace-sample/src/test/java/brave/internal/HexCodec.java:17]: I will remove this class altogether, since you've added it in `0.17.0`.

File: spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/trace/StackdriverTraceAutoConfiguration.java (hunk 2 above)
Comment (meltsufin) [reply spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/trace/StackdriverTraceAutoConfiguration.java:222]: Do you mean that the `B3Propagation.Factory` would be configured differently in zipkin-gcp, because we're not configuring it here?

**Comment by meltsufin** [comment]
@adriancole Can you please do a final review of this PR to make sure I configured the propagation factory correctly. Thanks!

File: spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/trace/StackdriverTraceAutoConfiguration.java (hunk 2 above)
Comment (adriancole) [reply spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/trace/StackdriverTraceAutoConfiguration.java]: I mean what we are overriding will be different configuration 2.2.x vs 3.0.x. Until we some annotation or otherwise to indicate the "real trace format" factory vs the wrapped one, this project will have to build differently for 2.2.x vs 3.0 or have some logic etc. @spencergibb mentioned maybe we can change 3.0 to use the `@Primary` annotation. Will think about it. Anyway, this is fine for 2.2.x

File: spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/trace/StackdriverTraceAutoConfiguration.java [hunk 4]
```diff
@@ -221,8 +219,9 @@ public Sender stackdriverSender(GcpTraceProperties traceProperties,
 
 	@Bean
 	@ConditionalOnMissingBean
-	public Propagation.Factory stackdriverPropagation() {
-		return StackdriverTracePropagation.FACTORY;
+	public BaggagePropagation.FactoryBuilder baggagePropagationFactoryBuilder() {
+		return BaggagePropagation.newFactoryBuilder(StackdriverTracePropagation.newFactory(
+				B3Propagation.newFactoryBuilder().injectFormat(B3Propagation.Format.MULTI).build()));
```
[the remark this answers is not in the document]
Comment (adriancole) [reply spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/trace/StackdriverTraceAutoConfiguration.java:224]: we don't have a way to pass this yet. so this is good enough for 2.2.x

[the remark this answers is not in the document]
Comment (meltsufin) [reply spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/trace/StackdriverTraceAutoConfiguration.java:224]: Thanks!
